### PR TITLE
Add "truncate_with_folder_marker" shorten strategy

### PR DESCRIPTION
**Pull request by dritter**
This adds @BenoitAverty `truncate_with_folder_marker` strategy for the `dir` segment.

### CODE REVIEW

File: powerlevel9k.zsh-theme [hunk 1]
```diff
@@ -610,6 +609,28 @@ prompt_dir() {
...
+            current_path="${current_path%/}/${marked_folder##*/}"
+          else
+            current_path="${current_path%/}/$POWERLEVEL9K_SHORTEN_DELIMITER/${marked_folder##*/}"
+          fi
+          last_marked_folder=$marked_folder
+        done
+
+        zero='%([BSUbfksu]|([FB]|){*})'
+        current_path=$current_path${PWD:${#${(S%%)last_marked_folder//$~zero/}}}
+      ;;
```
Comment (BenoitAverty) [inline powerlevel9k.zsh-theme]: I didn't really know what I was doing with this function. The "loop" technique works but I'm not sure what effet it has on performance. Maybe someone with a stronger knowledge of bash or zsh can optimize it later...

**Comment by bhilburn** [comment]
This is really well done, @dritter. I especially like your approach to creating user documentation, which I think makes things easier to use.

@BenoitAverty - This is a great idea, and I appreciate you providing input & review on this PR. Thanks for coming up with so many great improvements for the project!

File: powerlevel9k.zsh-theme (hunk 1 above)
Comment (bhilburn) [inline powerlevel9k.zsh-theme]: @dritter @BenoitAverty - What are your thoughts on Benoit's comment, here? Do you want to iterate on this function or should we merge as-is?

Comment (dritter) [inline powerlevel9k.zsh-theme]: @bhilburn Thanks for pointing me to this again. This is a quite complicated pattern to count the visible characters in the prompt. In our case we count the characters in `$last_marked_folder` (which should not contain any non-visible characters) and replace the amount of characters in `$PWD` from the left with nothing. IMHO this could be done way easier (just replace the characters from the left), but I think @BenoitAverty just took this pattern from the `truncate_with_package_name` strategy. As far as I can see this pattern originates from [barts prompt](https://opensource.apple.com/source/zsh/zsh-55/zsh/Functions/Prompts/prompt_bart_setup). For more information about this pattern, see [this answer](http://stackoverflow.com/a/40855342/5586433) on StackOverflow.

Regarding the performance of `upsearch` function. For me it looks quite solid. Nothing to optimize here.

**Comment by dritter** [comment]
I just got rid of the ugly pattern. Now the current path is truncated by removing the matched path from `upsearch` from `$PWD`. This is much more readable in my opinion.

@BenoitAverty could you confirm that everything still works as expected? I mean, I have tested it as well, the tests are green, but the more eyes, the better. 😉  Thanks in advance.

**Comment by dritter** [comment]
Btw. There is still this ugly pattern in `truncate_with_package_name`, I will fix this in a different PR.

**Comment by BenoitAverty** [comment]
I did a quick test just now, it seems to work OK. I may have forgotten some edge cases, but it works with one and two markers in the hierarchy so it should work in all cases.

**Comment by dritter** [comment]
Good to hear! Thanks for testing @BenoitAverty

**Comment by bhilburn** [comment]
Awesome :)
